### PR TITLE
[WIP] Get the wix tool working

### PR DESCRIPTION
**Pull request by mwichmann**
The wix tool uses the WiX Toolset for msi packaging.  It isn't
working, and is never run by scons' CI infrastructure so the problems
are not reported.

Add some code for better findability.

Problem #1: a language setting is required, and none of the tests do so.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [ ] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: src/engine/SCons/Tool/wix.py [hunk 1]
```diff
@@ -62,39 +64,58 @@ def generate(env):
...
-
+        if check_wix(env, path):
+            return 1
+
+    # Wix is not in scons' PATH. Let's look in typical install locations.
+    wix_env = os.environ.get('WIX', None)
+    if wix_env:
+        wix_installs = [os.path.join(wix_env, 'bin')]
+    else:
+        wix_path = r'C:\Program Files (x86)\WiX Toolset v*.*\bin'
```
Comment (dmoody256) [inline src/engine/SCons/Tool/wix.py:128]: This only would ever exist in "C:\Program Files (x86)" and not the 64 bit variant "C:\Program Files"?

File: src/engine/SCons/Tool/wix.py [hunk 2]
```diff
@@ -62,39 +64,58 @@ def generate(env):
...
     env['WIXLIGHT']  = 'light.exe'
 
-    # try to find the candle.exe and light.exe tools and 
-    # add the install directory to light libpath.
+    def check_wix(e, p):
+        """try to find the candle and light tools; add to path if found."""
+        try:
+            files = os.listdir(p)
+            if e['WIXCANDLE'] in files and e['WIXLIGHT'] in files:
+                e.PrependENVPath('PATH', p)
```
Comment (dmoody256) [inline src/engine/SCons/Tool/wix.py:95]: I took the "exists" function to mean "I want to know if this thing currently exists in my environment" not "find and add it to my environment". Is there any documentation regarding the goal of the exists function?

File: src/engine/SCons/Tool/wix.py [hunk 3]
```diff
@@ -62,39 +64,58 @@ def generate(env):
...
-                    env['WIXLIGHTFLAGS'] = []
-                return 1
-        except OSError:
-            pass # ignore this, could be a stale PATH entry.
-
+        if check_wix(env, path):
+            return 1
+
+    # Wix is not in scons' PATH. Let's look in typical install locations.
+    wix_env = os.environ.get('WIX', None)
```
Comment (dmoody256) [inline src/engine/SCons/Tool/wix.py:124]: I think the construction environment should tend not to be dependent on the system environment?

File: src/engine/SCons/Tool/wix.py (hunk 2 above)
Comment (mwichmann) [reply src/engine/SCons/Tool/wix.py:95]: I can't find any - I was asking this not long ago on the list (seems everybody just writes a tool by stealing from another tool, which is fine).  Yes, I expect you're right about the intent, this is the resurrection of an old patch before I'd seen some of the recent discussion.  I'm still willing to add some verbiage to the docs on `generate`/`exists` if I figure out what we should say.  
The problem is... if we find the tool and so report (only), then somebody else has to repeat the work later to find it again, and that's hardly ideal either?  We had that issue with the java tests until things got more unified.

File: src/engine/SCons/Tool/wix.py (hunk 1 above)
Comment (mwichmann) [reply src/engine/SCons/Tool/wix.py:128]: yes, apparently, the tools themeselves are 32-bit only (which I think is also true of all the visual studio stuff). No 64-bit installer exists.

**Comment by mwichmann** [comment]
So this did find wix, and did fail in the way the other PR did, but without forcing the install in appveror. I've closed the older PR, and we can use this one to decide whether further changes to the approach should be made. So far, haven't sorted where the problem comes from.  For reference, here is the error we get (all three msi tests show the same):
```
137/298 (45.97%) c:\\python36\\python.exe -tt test\packaging\msi\explicit-target.py
FAILED test of C:\projects\scons\src\script\scons.py
	at line 608 of C:\projects\scons\testing\framework\TestCommon.py (_complete)
	from line 711 of C:\projects\scons\testing\framework\TestCommon.py (run)
	from line 390 of C:\projects\scons\testing\framework\TestSCons.py (run)
	from line 76 of test\packaging\msi\explicit-target.py
C:\projects\scons\src\script\scons.py returned 2
STDOUT =========================================================================
scons: Reading SConscript files ...
STDERR =========================================================================
scons: *** Two environments with different actions were specified for the same target: mypackage.msi
File "C:\Users\appveyor\AppData\Local\Temp\1\testcmd.1988.bh44482n\SConstruct", line 19, in <module>
138/298 (46.31%) c:\\python36\\python.exe -tt test\packaging\msi\file-placement.py
FAILED test of C:\projects\scons\src\script\scons.py
```

**Comment by mwichmann** [comment]
Sigh. The linux build breaks because the wix test (not the msi packaging test) has a mock wix and the way it's found doesn't work (don't know why it worked before).  Should skip.

And the windows packaging tests are skipped because... they call whereis to decide whether to proceed or not, but do so in a context where packaging tool hasn't done the setup.

So I got no actual answer to me experiment.

File: src/engine/SCons/Tool/wix.py (hunk 1 above)
Comment (mwichmann) [reply src/engine/SCons/Tool/wix.py:128]: well, maybe... seems I've later been told if you do find a 32-bit windows install the path used by a installers will be "Program Files", not "Program Files (x86)"... I don't have one of those setups so I don't know if that applies to all packages, some packages, etc.

Comment (bdbaddog) [reply src/engine/SCons/Tool/wix.py:128]: Yes That's the case. We had to do this for finding vswhere.exe as well.
On a 32 bit install there is only Program files. On a 64 bit windows install there is Program Files (which is 64 bit), and Program Files (x86) which is 32 bit.

**Comment by mwichmann** [comment]
Is there any point in leaving this around, or shall I close and wait for someone to report needing it?

**Comment by mwichmann** [comment]
Withdrawn. We can bring this back from the dead if it becomes a problem.

**Pull request closed without merging**